pcap-linux.c

```diff
@@ -6900,6 +6922,70 @@ iface_get_offload(pcap_t *handle _U_)
 
 #endif /* HAVE_PF_PACKET_SOCKETS */
 
+static struct dsa_proto {
+	const char *name;
+	bpf_u_int32 linktype;
+} dsa_protos[] = {
+	{ "none", 0 },
+	{ "brcm", DLT_DSA_TAG_BRCM },
+	{ "brcm-prepend", DLT_DSA_TAG_BRCM_PREPEND },
+	{ "dsa", DLT_DSA_TAG_DSA },
+	{ "edsa", DLT_DSA_TAG_EDSA },
+	{ "gswip", DLT_DSA_TAG_GSWIP },
+	{ "ksz9477", DLT_DSA_TAG_KSZ9477 },
+	{ "lan9303", DLT_DSA_TAG_LAN9303 },
+	{ "mtk", DLT_DSA_TAG_MTK },
+	{ "qca", DLT_DSA_TAG_QCA },
+	{ "trailer", DLT_DSA_TAG_TRAILER },
+};
+
+static int	iface_dsa_get_proto_info(const char *device, pcap_t *handle,
+    char *ebuf)
+{
+	char *pathstr;
+	unsigned int i;
+	char buf[255];
+	ssize_t r;
+	int fd;
+
+	fd = asprintf(&pathstr, "/sys/class/net/%s/dsa/tagging", device);
+	if (fd < 0) {
+		pcap_fmt_errmsg_for_errno(handle->errbuf, PCAP_ERRBUF_SIZE,
+					  fd, "asprintf");
+		return PCAP_ERROR;
+	}
+
+	fd = open(pathstr, O_RDONLY);
+	free(pathstr);
+	/* This is not fatal, kernel >= 4.20 *might* expose this attribute */
+	if (fd < 0)
+		return 0;
+
+	memset(&buf, 0, sizeof(buf));
+	r = read(fd, buf, sizeof(buf));
+	close(fd);
+	if (r <= 0) {
+		pcap_fmt_errmsg_for_errno(handle->errbuf, PCAP_ERRBUF_SIZE,
+					  r, "read");
+		return PCAP_ERROR;
+	}
+
+	for (i = 0; i < sizeof(dsa_protos) / sizeof(dsa_protos[0]); i++) {
+		if (!strncmp(buf, dsa_protos[i].name,
```

Comment:
If the tag for the device is longer than the name in the table, this will do prefix matching; is that what it should do, or should it do an exact match?

If the tag for the device is *shorter* than the name in the table, this will run past the end of the tag string, as it hasn't been null-terminated.

You should probably:

- make the buffer 255+1 bytes, with the +1 for a terminating null character, read into the first 255 bytes, and set `buf[r]` to `'\0'`;
- before the `strncmp()` call, check whether `strlen(dsa_protos[i].name)` is equal to `r`, and then do a `strcmp()` call between `buf` and `dsa_protos[i].name` if the lengths are equal.